modules/category/category.service.js

```diff
@@ -22,39 +24,87 @@ class CategoryService {
     if (!categoryToUpdate) {
       throw new Error(CATEGORY_NOT_FOUND);
     }
-    if (await this.checkCategoryExist(category, id)) {
-      throw new Error(CATEGORY_ALREADY_EXIST);
-    }
     return await Category.findByIdAndUpdate(id, category, {
       new: true,
     });
   }
 
-  async addCategory(data) {
+  async addCategory(data, parentId) {
+    if (data.name.length < 2 || !data.code || (!data.isMain && !parentId)) {
+      throw new Error(WRONG_CATEGORY_DATA);
+    }
     if (await this.checkCategoryExist(data)) {
       throw new Error(CATEGORY_ALREADY_EXIST);
     }
-    return new Category(data).save();
+    let selectedId = parentId;
+    if (!parentId) {
+      selectedId = null;
+    }
+    const parentCategory = await Category.findOne({ _id: selectedId });
+    const newCategory = new Category(data);
+
+    if (selectedId && !parentCategory) {
+      throw new Error(CATEGORY_NOT_FOUND);
+    }
+    const savedCategory = await newCategory.save();
+    if (parentCategory) {
+      if (!parentCategory.isMain) {
+        throw new Error(CATEGORY_IS_NOT_MAIN);
+      }
+      if (data.isMain) {
+        throw new Error(WRONG_CATEGORY_DATA);
+      }
+      if (!parentCategory.available) {
+        savedCategory.available = false;
+      }
+      parentCategory.subcategories.push(savedCategory._id);
+      await parentCategory.save();
+    }
+    return await savedCategory.save();
   }
 
   async deleteCategory(id) {
     const category = await Category.findByIdAndDelete(id);
+
+    if (!category.isMain) {
+      const parentCategory = await Category.findOne({
+        subcategories: { $in: [id] },
+      });
+      await this.updateCategory(parentCategory._id, {
+        $pull: { subcategories: { $in: [id] } },
+      });
+    }
     if (category) {
       return category;
     }
     throw new Error(CATEGORY_NOT_FOUND);
   }
 
   async checkCategoryExist(data, id) {
+    if (!data.name.length) {
+      return false;
+    }
     const categoriesCount = await Category.countDocuments({
       _id: { $ne: id },
       name: {
         $elemMatch: {
-          $or: [{ value: data.name[0].value }, { value: data.name[1].value }],
+          $or: data.name.map(({ value }) => ({ value })),
         },
       },
     });
     return categoriesCount > 0;
   }
+
+  async getSubcategories(id) {
+    const category = await this.getCategoryById(id);
+    if (!category.isMain) {
+      return [];
+    }
+    const idList = category.subcategories.length
+      ? category.subcategories
+      : [''];
```

Comment:
Generally, It is an invalid value for the `$in` operator, so don't use it because it can cause really strange issues

Reply:
not providing an array will be crashing the operation, because it is required for $in operation to have an array